File: server/src/configs/firebase.ts
import admin, { ServiceAccount } from 'firebase-admin';
const serviceKey = process.env.FIREBASE_SERVICE_ACCOUNT_KEY
  ? JSON.parse(process.env.FIREBASE_SERVICE_ACCOUNT_KEY)
  : require('./serviceAccountKey.json');

if (process.env.NODE_ENV === 'test') {
  process.env.FIRESTORE_EMULATOR_HOST = 'localhost:8080';
  process.env.FIREBASE_AUTH_EMULATOR_HOST = 'localhost:9099';
  admin.initializeApp({
    projectId: 'weddingpass-dev'
  });
} else {
  admin.initializeApp({
    credential: admin.credential.cert(serviceKey as ServiceAccount),
  });
}

export const db = admin.firestore();
export const auth = admin.auth();
export const firebaseAdmin = admin;

File: server-go/internal/firebase/firebase.go
package firebase

import (
	"context"
	"log"
	"os"

	"cloud.google.com/go/firestore"
	firebase "firebase.google.com/go"
	"firebase.google.com/go/auth"
	"github.com/joho/godotenv"
	"google.golang.org/api/option"
)

var (
	FirebaseApp     *firebase.App
	FirestoreClient *firestore.Client
	AuthClient      *auth.Client
)

func InitFirebase(ctx context.Context) error {
	_ = godotenv.Load()

	nodeEnv := os.Getenv("ENV")
	var app *firebase.App
	var err error

	if nodeEnv == "test" {
		os.Setenv("FIRESTORE_EMULATOR_HOST", "localhost:8080")
		os.Setenv("FIREBASE_AUTH_EMULATOR_HOST", "localhost:9099")
		app, err = firebase.NewApp(ctx, &firebase.Config{
			ProjectID: "weddingpass-dev",
		})
		if err != nil {
			return err
		}
		log.Println("Firebase initialized in test mode")
	} else {
		serviceKey := os.Getenv("FIREBASE_SERVICE_ACCOUNT_KEY")
		var opts option.ClientOption
		if serviceKey != "" {
			opts = option.WithCredentialsJSON([]byte(serviceKey))
		} else {
			opts = option.WithCredentialsFile("../../internal/firebase/serviceAccountKey.json")
		}
		app, err = firebase.NewApp(ctx, nil, opts)
		if err != nil {
			return err
		}
		log.Println("Firebase initialized in production mode")
	}

	FirebaseApp = app

	fsClient, err := app.Firestore(ctx)
	if err != nil {
		return err
	}
	FirestoreClient = fsClient

	aClient, err := app.Auth(ctx)
	if err != nil {
		return err
	}
	AuthClient = aClient

	return nil
}
